lib/open_project/authentication/strategies/warden/basic_auth_failure.rb

```diff
@@ -0,0 +1,16 @@
+module OpenProject
+  module Authentication
+    module Strategies
+      module Warden
+        ##
+        # This strategy is inserted after optional basic auth strategies to
+        # indicate that invalid basic auth credentials were provided.
+        class BasicAuthFailure < ::Warden::Strategies::BasicAuth
+          def authenticate_user(_username, _password)
+            nil # always fails
```

Comment:
The `GlobalBasicAuth` and `UserBasicAuth` strategies are both optional. I.e. they are only triggered when certain user names are used for basic auth:

1) What ever is configured for the global basic auth access.
2) `apikey`

This means if a user provides a different user name the credentials would be wrong but the user would not get a 401 because the strategies were not triggered due to the different user name. Instead the session strategy will be used as a fallback and the user will be either logged in or anonymous.

But when they provide credentials they should get a 401 if they are wrong. This is what this strategy is for.
It always fails. But is only triggered when basic auth credentials are provided. If the credentials are correct, one of the previous two basic auth strategies will succeed. If not they must be wrong.

This wouldn't be necessary if the two basic auth strategies were always triggered. For instance if the `UserBasicAuth` strategy would work with usernames and passwords instead of `apikey`.